Confirmed, and fixed in 77a87354.

Reproduced it exactly. Both in-tree loops derive from `AbstractSelectorIOLoop`, so `getattr(loop_class, name)` resolves a method the loop never defines through the MRO to the protocol's own body. Adding a member with a body that no loop implements passed vacuously for `select_connection.IOLoop` and `_GeventSelectorIOLoop`, and failed only for tornado's `IOLoop`, which does not derive. So "immediately required of all three loops" held for tornado alone, exactly as your `defined_by` output shows. And with the decorators gone, an inherited stub is the failure the test is supposed to catch: a forgotten `remove_handler` would resolve to the do-nothing body rather than raising.

I took your suggested check: walk `__mro__` for the class that defines each name and reject `AbstractSelectorIOLoop` itself, so an inherited stub fails the same as an absent method.

To confirm it bites, I added a bodied member to the protocol that no loop implements. Under the old test that failed on tornado only; under the new one it fails on all three:

```console
$ hatch run pytest tests/unit/selector_ioloop_adapter_tests.py -n 0 -q
FAILED ...::SelectConnectionIOLoopConformanceTests::test_implements_every_protocol_method
FAILED ...::GeventSelectorIOLoopConformanceTests::test_implements_every_protocol_method
FAILED ...::TornadoIOLoopConformanceTests::test_implements_every_protocol_method
FAILED ...::AbstractSelectorIOLoopTests::test_isinstance_accepts_a_loop_that_does_not_derive_from_it
4 failed, 7 passed
```

With the mutation reverted, at the pushed tip: `lint-check`, `fmt-check`, `docfmt-check`, `typecheck` all exit 0; `hatch run test` 1473 passed, 61 skipped against a live broker; 3.7.17 unit suite 912 passed, 24 skipped, the 2 `isinstance` tests skipping by design.

--- a/pika/_utils.py
+++ b/pika/_utils.py
@@ -25,7 +25,25 @@ else:
         return func
 
 
-__all__ = ['override']
+# `Protocol` and `runtime_checkable` (PEP 544) landed in the stdlib in 3.8,
+# while pika still supports 3.7. Same dependency-free approach as `override`
+# above: type checkers run at 3.10 or later and see the real objects, while a
+# 3.7 runtime falls back to a plain base class and a no-op decorator. Protocols
+# in pika are only ever subclassed explicitly, never instantiated or passed to
+# `isinstance`, so the nominal fallback behaves identically at runtime.
+# Re-exported across pika (declared in __all__ below).
+if sys.version_info >= (3, 8):
+    from typing import Protocol, runtime_checkable
+elif TYPE_CHECKING:
+    from typing_extensions import Protocol, runtime_checkable
+else:
+    Protocol = object
+
+    def runtime_checkable(cls):
+        return cls
+
+
+__all__ = ['Protocol', 'override', 'runtime_checkable']
 
 RE_NUM = re.compile(r'(\d+).+')
 

--- a/pika/adapters/gevent_connection.py
+++ b/pika/adapters/gevent_connection.py
@@ -212,9 +212,9 @@ class _GeventSelectorIOLoop(AbstractSelectorIOLoop):
     # Gevent's READ and WRITE masks are defined as 1 and 2 respectively. No
     # ERROR mask is defined.
     # See https://www.gevent.org/api/gevent.hub.html#gevent._interfaces.ILoop.io
-    READ = 1  # pyright: ignore[reportAssignmentType, reportIncompatibleMethodOverride]
-    WRITE = 2  # pyright: ignore[reportAssignmentType, reportIncompatibleMethodOverride]
-    ERROR = 0  # pyright: ignore[reportAssignmentType, reportIncompatibleMethodOverride]
+    READ = 1
+    WRITE = 2
+    ERROR = 0
 
     def __init__(self, gevent_hub: gevent.hub.Hub | None = None) -> None:
         """

--- a/pika/adapters/select_connection.py
+++ b/pika/adapters/select_connection.py
@@ -403,9 +403,9 @@ class IOLoop(AbstractSelectorIOLoop):
     """
 
     # READ/WRITE/ERROR per `AbstractSelectorIOLoop` requirements
-    READ = PollEvents.READ  # pyright: ignore[reportIncompatibleMethodOverride, reportAssignmentType]
-    WRITE = PollEvents.WRITE  # pyright: ignore[reportIncompatibleMethodOverride, reportAssignmentType]
-    ERROR = PollEvents.ERROR  # pyright: ignore[reportIncompatibleMethodOverride, reportAssignmentType]
+    READ = PollEvents.READ
+    WRITE = PollEvents.WRITE
+    ERROR = PollEvents.ERROR
 
     def __init__(self) -> None:
         self._timer = _Timer()

--- a/pika/adapters/tornado_connection.py
+++ b/pika/adapters/tornado_connection.py
@@ -69,8 +69,7 @@ class TornadoConnection(base_connection.BaseConnection):
             nbio = custom_ioloop
         else:
             nbio = (selector_ioloop_adapter.SelectorIOServicesAdapter(
-                custom_ioloop or
-                ioloop.IOLoop.instance()))  # type: ignore[arg-type]
+                custom_ioloop or ioloop.IOLoop.instance()))
         super().__init__(
             parameters,
             on_open_callback,
@@ -98,7 +97,7 @@ class TornadoConnection(base_connection.BaseConnection):
         :param workflow: Optional connection workflow instance to use
         """
         nbio = selector_ioloop_adapter.SelectorIOServicesAdapter(
-            custom_ioloop or ioloop.IOLoop.instance())  # type: ignore[arg-type]
+            custom_ioloop or ioloop.IOLoop.instance())
 
         def connection_factory(
                 params: connection.Parameters | None) -> TornadoConnection:

--- a/pika/adapters/utils/selector_ioloop_adapter.py
+++ b/pika/adapters/utils/selector_ioloop_adapter.py
@@ -7,9 +7,9 @@ from __future__ import annotations
 import logging
 import socket
 import threading
-from typing import Any, Callable, Protocol, runtime_checkable
+from typing import Any, Callable
 
-from pika._utils import override
+from pika._utils import Protocol, override, runtime_checkable
 from pika.adapters.utils import io_services_utils, nbio_interface
 from pika.adapters.utils.io_services_utils import check_callback_arg, check_fd_arg
 
@@ -28,7 +28,8 @@ class AbstractSelectorIOLoop(Protocol):
 
     A `Protocol` rather than a base class so that loops implementing this
     interface without deriving from it, such as `tornado.ioloop.IOLoop`, satisfy
-    it structurally.
+    it structurally. `Protocol` reached the stdlib in 3.8, so on 3.7 this falls
+    back to a plain base class; see `pika._utils`.
     """
 
     # The values of the I/O loop's READ/WRITE/ERROR flags, which may be used

--- a/pika/adapters/utils/selector_ioloop_adapter.py
+++ b/pika/adapters/utils/selector_ioloop_adapter.py
@@ -4,11 +4,10 @@ as tornado's and our home-grown select_connection's I/O loops.
 
 from __future__ import annotations
 
-import abc
 import logging
 import socket
 import threading
-from typing import Any, Callable
+from typing import Any, Callable, Protocol, runtime_checkable
 
 from pika._utils import override
 from pika.adapters.utils import io_services_utils, nbio_interface
@@ -17,7 +16,8 @@ from pika.adapters.utils.io_services_utils import check_callback_arg, check_fd_a
 LOGGER = logging.getLogger(__name__)
 
 
-class AbstractSelectorIOLoop:
+@runtime_checkable
+class AbstractSelectorIOLoop(Protocol):
     """
     Selector-based I/O loop interface expected by
     `selector_ioloop_adapter.SelectorIOServicesAdapter`
@@ -25,34 +25,19 @@ class AbstractSelectorIOLoop:
     NOTE: this interface follows the corresponding methods and attributes
      of `tornado.ioloop.IOLoop` in order to avoid additional adapter layering
      when wrapping tornado's IOLoop.
+
+    A `Protocol` rather than a base class so that loops implementing this
+    interface without deriving from it, such as `tornado.ioloop.IOLoop`, satisfy
+    it structurally.
     """
 
-    @property
-    @abc.abstractmethod
-    def READ(self) -> int:
-        """
-        The value of the I/O loop's READ flag; READ/WRITE/ERROR may be used with bitwise operators
-        as expected.
+    # The values of the I/O loop's READ/WRITE/ERROR flags, which may be used
+    # with bitwise operators as expected. Declared as attributes rather than
+    # properties so that implementations may use either.
+    READ: int
+    WRITE: int
+    ERROR: int
 
-        Implementation note: the implementations can simply replace these
-        READ/WRITE/ERROR properties with class-level attributes
-        """
-
-    @property
-    @abc.abstractmethod
-    def WRITE(self) -> int:
-        """The value of the I/O loop's WRITE flag; READ/WRITE/ERROR may be used with bitwise
-        operators as expected.
-        """
-
-    @property
-    @abc.abstractmethod
-    def ERROR(self) -> int:
-        """The value of the I/O loop's ERROR flag; READ/WRITE/ERROR may be used with bitwise
-        operators as expected.
-        """
-
-    @abc.abstractmethod
     def close(self) -> None:
         """
         Release IOLoop's resources.
@@ -62,7 +47,6 @@ class AbstractSelectorIOLoop:
         `IOLoop` should be performed.
         """
 
-    @abc.abstractmethod
     def start(self) -> None:
         """
         Run the I/O loop.
@@ -70,7 +54,6 @@ class AbstractSelectorIOLoop:
         It will loop until requested to exit. See `stop()`.
         """
 
-    @abc.abstractmethod
     def stop(self) -> None:
         """
         Request exit from the ioloop.
@@ -83,7 +66,6 @@ class AbstractSelectorIOLoop:
         `ioloop.add_callback(ioloop.stop)`
         """
 
-    @abc.abstractmethod
     def call_later(self, delay: float, callback: Callable[..., Any]) -> object:
         """
         Add the callback to the IOLoop timer to be called after delay seconds from the time of call
@@ -94,7 +76,6 @@ class AbstractSelectorIOLoop:
         :returns: handle to the created timeout that may be passed to `remove_timeout()`
         """
 
-    @abc.abstractmethod
     def remove_timeout(self, timeout_handle: Any) -> None:
         """
         Remove a timeout.
@@ -102,7 +83,6 @@ class AbstractSelectorIOLoop:
         :param timeout_handle: Handle of timeout to remove
         """
 
-    @abc.abstractmethod
     def add_callback(self, callback: Callable[..., Any]) -> None:
         """
         Requests a call to the given function as soon as possible in the context of this IOLoop's
@@ -118,7 +98,6 @@ class AbstractSelectorIOLoop:
         :param callback: The callback method
         """
 
-    @abc.abstractmethod
     def add_handler(self, fd: int, handler: Callable[[int, int], None],
                     events: int) -> None:
         """
@@ -129,7 +108,6 @@ class AbstractSelectorIOLoop:
         :param events: The event mask using READ, WRITE, ERROR.
         """
 
-    @abc.abstractmethod
     def update_handler(self, fd: int, events: int) -> None:
         """
         Changes the events we watch for.
@@ -138,7 +116,6 @@ class AbstractSelectorIOLoop:
         :param events: The event mask using READ, WRITE, ERROR
         """
 
-    @abc.abstractmethod
     def remove_handler(self, fd: int) -> None:
         """
         Stop watching the given file descriptor for events.

--- a/pika/adapters/utils/selector_ioloop_adapter.py
+++ b/pika/adapters/utils/selector_ioloop_adapter.py
@@ -30,11 +30,22 @@ class AbstractSelectorIOLoop(Protocol):
     interface without deriving from it, such as `tornado.ioloop.IOLoop`, satisfy
     it structurally. `Protocol` reached the stdlib in 3.8, so on 3.7 this falls
     back to a plain base class; see `pika._utils`.
+
+    The methods carry no `@abc.abstractmethod` decorators. `typing._ProtocolMeta`
+    derives from `abc.ABCMeta`, so adding them back would populate
+    `__abstractmethods__` and make an incomplete subclass fail at construction on
+    3.8 and later, but do nothing at all on 3.7, where the base is `object`.
+    mypy reports a missing method either way, so the decorators would buy
+    version-dependent runtime behavior and nothing else.
     """
 
     # The values of the I/O loop's READ/WRITE/ERROR flags, which may be used
-    # with bitwise operators as expected. Declared as attributes rather than
-    # properties so that implementations may use either.
+    # with bitwise operators as expected. Protocol variable members are not
+    # satisfied by read-only properties, so implementations must expose plain
+    # `int` attributes. Declaring these as properties here instead would accept
+    # both shapes under mypy, but pyright then rejects the `int` class
+    # attributes both in-tree loops use, which is what the six
+    # `pyright: ignore` comments this change retires were suppressing.
     READ: int
     WRITE: int
     ERROR: int

--- a/tests/unit/selector_ioloop_adapter_tests.py
+++ b/tests/unit/selector_ioloop_adapter_tests.py
@@ -1,0 +1,160 @@
+"""Tests for pika.adapters.utils.selector_ioloop_adapter.AbstractSelectorIOLoop."""
+
+import inspect
+import sys
+import unittest
+
+import pika._utils
+from pika.adapters import select_connection
+from pika.adapters.utils.selector_ioloop_adapter import AbstractSelectorIOLoop
+
+try:
+    import tornado.ioloop
+    HAS_TORNADO = True
+except ImportError:
+    HAS_TORNADO = False
+
+try:
+    from pika.adapters.gevent_connection import _GeventSelectorIOLoop
+    HAS_GEVENT = True
+except ImportError:
+    _GeventSelectorIOLoop = None
+    HAS_GEVENT = False
+
+FLAG_NAMES = ('READ', 'WRITE', 'ERROR')
+
+# Derived from the protocol itself rather than hard-coded, so a method added to
+# the interface is immediately required of every loop checked below.
+PROTOCOL_METHOD_NAMES = tuple(
+    sorted(name for name, value in vars(AbstractSelectorIOLoop).items()
+           if not name.startswith('_') and callable(value)))
+
+
+class _StructuralLoop:
+    """Implements the whole interface without deriving from it, as tornado's `IOLoop` does."""
+
+    READ = 1
+    WRITE = 2
+    ERROR = 4
+
+    def close(self):
+        pass
+
+    def start(self):
+        pass
+
+    def stop(self):
+        pass
+
+    def call_later(self, delay, callback):
+        pass
+
+    def remove_timeout(self, timeout_handle):
+        pass
+
+    def add_callback(self, callback):
+        pass
+
+    def add_handler(self, fd, handler, events):
+        pass
+
+    def update_handler(self, fd, events):
+        pass
+
+    def remove_handler(self, fd):
+        pass
+
+
+class _IncompleteLoop:
+    """Carries the flags but none of the methods."""
+
+    READ = 1
+    WRITE = 2
+    ERROR = 4
+
+
+class _ConformanceChecks:
+    """
+    Checks that a loop class satisfies `AbstractSelectorIOLoop`.
+
+    Mixed into a `unittest.TestCase` per loop class. Everything is checked against the class rather
+    than an instance so that no loop, hub, or file descriptor is created.
+    """
+
+    loop_class = None
+
+    def test_flags_are_int_attributes(self):
+        # The protocol declares READ/WRITE/ERROR as variable members, which a
+        # read-only property does not satisfy. `getattr_static` is what
+        # distinguishes the two: for a property it returns the property object
+        # rather than the value.
+        for name in FLAG_NAMES:
+            value = inspect.getattr_static(self.loop_class, name)
+            self.assertIsInstance(value, int, f'{name} is not a plain int')
+
+    def test_implements_every_protocol_method(self):
+        self.assertTrue(PROTOCOL_METHOD_NAMES)
+        for name in PROTOCOL_METHOD_NAMES:
+            self.assertTrue(callable(getattr(self.loop_class, name, None)),
+                            f'{name} is missing or not callable')
+
+
+class SelectConnectionIOLoopConformanceTests(_ConformanceChecks,
+                                             unittest.TestCase):
+    loop_class = select_connection.IOLoop
+
+
+@unittest.skipUnless(HAS_GEVENT, 'gevent not installed')
+@unittest.skipIf(pika._utils.ON_WINDOWS, 'Windows not supported')
+class GeventSelectorIOLoopConformanceTests(_ConformanceChecks,
+                                           unittest.TestCase):
+    loop_class = _GeventSelectorIOLoop
+
+
+@unittest.skipUnless(HAS_TORNADO, 'tornado not installed')
+class TornadoIOLoopConformanceTests(_ConformanceChecks, unittest.TestCase):
+    """
+    Tornado's `IOLoop` is the reason the interface is structural.
+
+    It implements every member without deriving from the protocol, which is what allowed the two
+    `[arg-type]` suppressions in `tornado_connection.py` to be removed.
+    """
+
+    loop_class = tornado.ioloop.IOLoop if HAS_TORNADO else None
+
+    def test_does_not_derive_from_the_protocol(self):
+        self.assertNotIn(AbstractSelectorIOLoop, self.loop_class.__mro__)
+
+
+class AbstractSelectorIOLoopTests(unittest.TestCase):
+
+    def test_flags_are_declared_as_variable_members(self):
+        # Declaring the flags as properties instead would let a loop expose them
+        # as read-only properties, but pyright then rejects the plain `int` class
+        # attributes both in-tree loops use. A property declaration puts a
+        # property object in the class dict; an annotation does not.
+        self.assertEqual(set(AbstractSelectorIOLoop.__annotations__),
+                         set(FLAG_NAMES))
+        for name in FLAG_NAMES:
+            self.assertNotIn(name, vars(AbstractSelectorIOLoop))
+
+    def test_carries_no_abstract_methods(self):
+        # `typing._ProtocolMeta` derives from `abc.ABCMeta`, so an
+        # `@abc.abstractmethod` here would block construction of an incomplete
+        # subclass on 3.8 and later while doing nothing at all on 3.7, where the
+        # protocol falls back to `object`.
+        self.assertEqual(
+            getattr(AbstractSelectorIOLoop, '__abstractmethods__', frozenset()),
+            frozenset())
+
+    @unittest.skipIf(sys.version_info < (3, 8),
+                     '`Protocol` falls back to a nominal base class on 3.7')
+    def test_isinstance_accepts_a_loop_that_does_not_derive_from_it(self):
+        loop = _StructuralLoop()
+        self.assertNotIn(AbstractSelectorIOLoop, type(loop).__mro__)
+        self.assertIsInstance(loop, AbstractSelectorIOLoop)
+
+    @unittest.skipIf(sys.version_info < (3, 8),
+                     '`Protocol` falls back to a nominal base class on 3.7')
+    def test_isinstance_rejects_an_incomplete_loop(self):
+        self.assertNotIsInstance(_IncompleteLoop(), AbstractSelectorIOLoop)

